File: mcmt-tracking/mcmt-tracking/src/mcmt_track/src/mcmt_track_utils.cpp
/**
 * @file mcmt_track_utils.cpp
 * @author your name (you@domain.com)
 * @author Niven Sie, sieniven@gmail.com
 * @author Seah Shao Xuan
 * 
 * This code contains the common functions and classes CameraTracks and TrackPlot
 * that are used in our main tracking and re-identifcation algorithm.
 */

#define _USE_MATH_DEFINES

#include <mcmt_track/mcmt_track_utils.hpp>
#include <stdlib.h>
#include <math.h>
#include <memory>
#include <algorithm>
#include <functional>

using namespace mcmt;

/**
 * This class is for tracking the camera's track plots
 */
CameraTracks::CameraTracks(int index)
{
	// set camera index
	index_ = index;
}

/**
 * This class is for storing our tracks' information, calculating track feature
 * variable, and caluclating geometrical angles and distances of the tracks.
 */
TrackPlot::TrackPlot(int track_id)
{
	// set the trackid and the mismatch count to initial value of zero
	id_ = track_id;
	mismatch_count_ = 0;
	lastSeen_ = 0;
}

/**
 * This function updates the track with the latest track information
 */
void TrackPlot::update(std::vector<int> & location, int & size, int & frame_no)
{
	xs_.push_back(location[0]);
	ys_.push_back(location[1]);
	frameNos_.push_back(frame_no);
	size_.push_back(size);
	lastSeen_ = frame_no;
}

/**
 * This function calculates the track feature variable of the track in the 
 * current frame, and stores the value of the track feature variable inside
 * the vector track_feature_variable_
 */
void TrackPlot::calculate_track_feature_variable(int & frame_no, int & fps)
{
	// checks if there is enough data to calculate the turning angle (at least
	// 3 points) and that the data is from the current frame
	if (frameNos_.size() >= 3 && frameNos_.back() == frame_no) {
		// check if the last 3 frames are consecutive
		if ((frameNos_.end()[-2] == (frameNos_.end()[-1] - 1)) &&
				(frameNos_.end()[-3] == (frameNos_.end()[-2] - 1))) {
			// retrieve the x and y values of the last 3 points
			int t = 2;
			std::vector<int> x(xs_.end() - 3, xs_.end());
			std::vector<int> y(ys_.end() - 3, ys_.end());

			// Turning angle
			// ~this code block has not been ported yet as the old method of calculation
			// is not in the current pipeline. to see the old methodology, refer to the
			// python version of mcmt_track_utils.py

			// Curvature
			// ~this code block has not been ported yet as the old method of calculation
			// is not in the current pipeline. to see the old methodology, refer to the
			// python version of mcmt_track_utils.py

			// Pace
			// using dt = 1/fps
			float d = hypot((x[t] - x[t - 1]), (y[t] - y[t - 1]));
			float pace = d * fps;
			pace_.push_back(pace);

			// append track feature variable. the current pipeline uses pace as our
			// track feature variable value
			track_feature_variable_.push_back(pace);
		}
	}
}

/**
 * This function checks if the current TrackPlot has been stationary for too long,
 * by calculating the distance the track moved for the last 9 frames. it returns
 * a bool True if its average distance moved per frame is less than 3.0, and returns
 * False if otherwise
 */
bool TrackPlot::check_stationary()
{
	// declare function variable to store total Euclidean distance traveled
	float distance = 0.0;

	for (int i = -1; i >= -9; i--) {
		int dx = xs_[xs_.size() + i] - xs_[xs_.size() + i - 1];
		int dy = ys_[ys_.size() + i] - ys_[ys_.size() + i - 1];

		// get euclidean distance, and add to total distance
		distance += hypot(dx, dy);
	}

	distance /= 9;

	if (distance < 3.0) {
		return true;
	} else {
		return false;
	}
}

/**
 * this function updates the current frame's other_tracks list
 */
void mcmt::update_other_tracks(std::shared_ptr<mcmt::TrackPlot> trackplot,
	std::shared_ptr<CameraTracks> & cumulative_track)
{
	// clear other_tracks vector
	trackplot->other_tracks_.clear();

	// iterate through the camera's tracks
	std::map<int, std::shared_ptr<mcmt::TrackPlot>>::iterator other_track;
	for (other_track = cumulative_track->track_plots_.begin(); 
		other_track != cumulative_track->track_plots_.end(); other_track++) {

		if (other_track->second->xs_.size() != 0 && other_track->second->ys_.size() != 0) {
			int dx = other_track->second->xs_.end()[-1] - trackplot->xs_.end()[-1];
			int dy = other_track->second->ys_.end()[-1] - trackplot->ys_.end()[-1];

			if (dx != 0 && dy != 0) {
				auto new_other_track = std::shared_ptr<TrackPlot::OtherTrack>(new TrackPlot::OtherTrack());
				new_other_track->angle = atan2(dy, dx);
				new_other_track->dist = hypot(dx, dy);
				trackplot->other_tracks_.push_back(new_other_track);
			}
		}
	}

	for (other_track = cumulative_track->track_new_plots_.begin(); 
		other_track != cumulative_track->track_new_plots_.end(); other_track++) {
		if (other_track->second->xs_.size() != 0 && other_track->second->ys_.size() != 0) {
			int dx = other_track->second->xs_.end()[-1] - trackplot->xs_.end()[-1];
			int dy = other_track->second->ys_.end()[-1] - trackplot->ys_.end()[-1];

			if (dx != 0 && dy != 0) {
				auto new_other_track = std::shared_ptr<TrackPlot::OtherTrack>(new TrackPlot::OtherTrack());
				new_other_track->angle = atan2(dy, dx);
				new_other_track->dist = hypot(dx, dy);
				trackplot->other_tracks_.push_back(new_other_track);
			}
		}
	}
}

void mcmt::combine_track_plots(
	int & index,
	std::shared_ptr<CameraTracks> camera_tracks,
	std::shared_ptr<mcmt::TrackPlot> track_plot,
	int & frame_count)
{
	// append the various variable value vectors
	camera_tracks->track_plots_[index]->xs_.insert(
		camera_tracks->track_plots_[index]->xs_.end(),
		track_plot->xs_.begin(),
		track_plot->xs_.end());
	
	camera_tracks->track_plots_[index]->ys_.insert(
		camera_tracks->track_plots_[index]->ys_.end(),
		track_plot->ys_.begin(),
		track_plot->ys_.end());
	
	camera_tracks->track_plots_[index]->frameNos_.insert(
		camera_tracks->track_plots_[index]->frameNos_.end(),
		track_plot->frameNos_.begin(),
		track_plot->frameNos_.end());

	camera_tracks->track_plots_[index]->turning_angle_.insert(
		camera_tracks->track_plots_[index]->turning_angle_.end(),
		track_plot->turning_angle_.begin(),
		track_plot->turning_angle_.end());

	camera_tracks->track_plots_[index]->curvature_.insert(
		camera_tracks->track_plots_[index]->curvature_.end(),
		track_plot->curvature_.begin(),
		track_plot->curvature_.end());

	camera_tracks->track_plots_[index]->pace_.insert(
		camera_tracks->track_plots_[index]->pace_.end(),
		track_plot->pace_.begin(),
		track_plot->pace_.end());

	camera_tracks->track_plots_[index]->track_feature_variable_.insert(
		camera_tracks->track_plots_[index]->track_feature_variable_.end(),
		track_plot->track_feature_variable_.begin(),
		track_plot->track_feature_variable_.end());

	camera_tracks->track_plots_[index]->lastSeen_ = frame_count;
}

/**
 * this function converts scalar value into rbg values
 */
std::vector<int> mcmt::scalar_to_rgb(int & scalar_value, int & max_value)
{
	// create empty vector
	std::vector<int> output;

	double a = (1.0 - ((double) scalar_value / (double) max_value)) * 5.0;
	int x = int(floor(a));
	int y = int(floor(255 * (a - (double) x)));

	if (x == 0) {
		output.push_back(255);
		output.push_back(y);
		output.push_back(0);
	} else if (x == 1) {
		output.push_back(255);
		output.push_back(255);
		output.push_back(0);
	} else if (x == 2) {
		output.push_back(0);
		output.push_back(255);
		output.push_back(y);
	} else if (x == 3) {
		output.push_back(0);
		output.push_back(255);
		output.push_back(255);
	} else if (x == 4) {
		output.push_back(y);
		output.push_back(0);
		output.push_back(255);
	} else {	// x = 5
		output.push_back(255);
		output.push_back(0);
		output.push_back(255);
	}
	return output;
}

File: mcmt-tracking/src/mcmt_track/mcmt_track/mcmt_track_utils.py
# basic Python frameworks
import math
import numpy as np


class CameraTracks:
    def __init__(self, index):
        self.index = index
        self.track_plots_ids = []
        self.track_new_plots_ids = []
        
        # self.track_plots = []
        # self.track_new_plots = []
        self.track_plots = {}
        self.track_new_plots = {}

        self.output_log = []


class TrackPlot:
    def __init__(self, track_id):
        self.id = track_id
        self.xs = np.array([], dtype=int)
        self.ys = np.array([], dtype=int)
        self.frameNos = np.array([], dtype=int)
        self.times = np.array([])
        self.colourized_times = []
        self.lastSeen = 0

        self.Q = 0
        self.turning_angle = np.array([])  # Degrees
        self.curvature = np.array([])
        self.pace = np.array([])
        self.track_feature_variable = np.array([])
        self.other_tracks = []

        self.xyz = None


    def update(self, location, frame_no, time=None):
        self.xs = np.append(self.xs, [int(location[0])])
        self.ys = np.append(self.ys, [int(location[1])])
        self.frameNos = np.append(self.frameNos, [frame_no])

        if time is not None:
            self.times = np.append(self.times, [time])

        self.lastSeen = frame_no


    def calculate_track_feature_variable(self, frameNo, fps):
        # Check if there is enough data to calculate turning angle (at least 3 points)
        # And that the data is still current
        if len(self.frameNos) >= 3 and self.frameNos[-1] == frameNo:
            # Check if the last 3 frames are consecutive
            if self.frameNos[-2] == self.frameNos[-1] - 1 and self.frameNos[-3] == self.frameNos[-2] - 1:
                # Retrieve the x and y values of the last 3 points and introduce t for readability
                t = 2
                x, y = self.xs[-3:], self.ys[-3:]

                # Turning angle
            #    xs = np.array([x[t] - x[t - 1], x[t - 1] - x[t - 2]])
            #    ys = np.array([y[t] - y[t - 1], y[t - 1] - y[t - 2]])

                # arctan2 returns the element-wise arc tangent, choosing the element correctly
                # Special angles excluding infinities:
                # y = +/- 0, x = +0, theta = +/- 0
                # y = +/- 0, x = -0, theta = +/- pi
                # whats a positive or negative 0?
            #    heading = np.arctan2(ys, xs) * 180 / np.pi

            #    turning_angle = heading[1] - heading[0]

            #    self.turning_angle = np.append(self.turning_angle, turning_angle)

                # Curvature
            #    a = np.sqrt((x[t] - x[t - 2]) ** 2 + (y[t] - y[t - 2]) ** 2)
            #    b = np.sqrt((x[t - 1] - x[t - 2]) ** 2 + (y[t - 1] - y[t - 2]) ** 2)
                c = np.sqrt((x[t] - x[t - 1]) ** 2 + (y[t] - y[t - 1]) ** 2)

            #    if b == 0 or c == 0:
            #        curvature = 0
            #    else:
            #       if abs((a ** 2 - b ** 2 - c ** 2) / (2 * b * c)) <= 1:
            #            curvature = np.arccos((a ** 2 - b ** 2 - c ** 2) / (2 * b * c))
                        # For whatever reason, the arccos of 1.0000000000000002 is nan
            #        else:
            #            curvature = 0
            #        if np.isnan(curvature):
            #            curvature = 0

            #    self.curvature = np.append(self.curvature, curvature)

                # Pace
                # Check if the data was returned in real time
                if self.times.size != 0:  # If so, dt is the difference in the time each consecutive frame was read
                    dt = self.times[-1] - self.times[-2]
                else:
                    # assume 30 FPS
                    dt = 1 / fps

                pace = c / dt
                self.pace = np.append(self.pace, pace)

                # track_feature_variable = np.mean(self.turning_angle) * np.mean(self.curvature) * np.mean(self.pace)
                # self.track_feature_variable = np.append(self.track_feature_variable, track_feature_variable)
                self.track_feature_variable = np.append(self.track_feature_variable, pace)


    def update_other_tracks(self, cumulative_tracks):
        
        self.other_tracks = []

        for other_track in list(cumulative_tracks.track_new_plots.values()) + list(cumulative_tracks.track_plots.values()):

            if len(other_track.xs) != 0 and len(other_track.ys) != 0: 
                dx = other_track.xs[-1] - self.xs[-1]
                dy = other_track.ys[-1] - self.ys[-1]

                if dx != 0 and dy!= 0:
                    self.other_tracks.append((np.arctan2(dy, dx), math.hypot(dx, dy)))


def combine_track_plots(index, camera_tracks, track_plot, frame_count):
    # appending of various variable value lists
    camera_tracks.track_plots[index].xs = np.append(camera_tracks.track_plots[index].xs, track_plot.xs)
    camera_tracks.track_plots[index].ys = np.append(camera_tracks.track_plots[index].ys, track_plot.ys)
    camera_tracks.track_plots[index].frameNos = np.append(camera_tracks.track_plots[index].frameNos, track_plot.frameNos)
    camera_tracks.track_plots[index].times = np.append(camera_tracks.track_plots[index].times, track_plot.times)
    camera_tracks.track_plots[index].colourized_times = np.append(camera_tracks.track_plots[index].colourized_times,
                                                                  track_plot.colourized_times)
    camera_tracks.track_plots[index].lastSeen = frame_count
    camera_tracks.track_plots[index].turning_angle = np.append(camera_tracks.track_plots[index].turning_angle,
                                                               track_plot.turning_angle)
    camera_tracks.track_plots[index].curvature = np.append(camera_tracks.track_plots[index].xs, track_plot.xs)
    camera_tracks.track_plots[index].pace = np.append(camera_tracks.track_plots[index].pace, track_plot.pace)
    camera_tracks.track_plots[index].track_feature_variable = np.append(camera_tracks.track_plots[index].track_feature_variable,
                                                                        track_plot.track_feature_variable)


def scalar_to_rgb(scalar_value, max_value):
    f = scalar_value / max_value
    a = (1 - f) * 5
    x = math.floor(a)
    y = math.floor(255 * (a - x))
    if x == 0:
        return 255, y, 0
    elif x == 1:
        return 255, 255, 0
    elif x == 2:
        return 0, 255, y
    elif x == 3:
        return 0, 255, 255
    elif x == 4:
        return y, 0, 255
    else:  # x == 5:
        return 255, 0, 255